mm2src/lp_price.rs

```diff
@@ -0,0 +1,336 @@
+                    };
+                rate_infos.base_price = base_price_infos.last_price.clone();
+                rate_infos.rel_price = rel_price_infos.last_price.clone();
+                rate_infos.price = &base_price_infos.last_price / &rel_price_infos.last_price;
+                Some(rate_infos)
+            },
+            None => None,
+        }
+    }
+}
+
+pub async fn process_price_request(price_url: &str) -> Result<TickerInfosRegistry, MmError<PriceServiceRequestError>> {
+    debug!("Fetching price from: {}", price_url);
+    let (status, headers, body) = slurp_url(price_url).await?;
+    let (status_code, body, _) = (status, std::str::from_utf8(&body)?.trim().into(), headers);
+    if status_code != StatusCode::OK {
+        return MmError::err(PriceServiceRequestError::HttpProcessError(body));
+    }
+    let model: HashMap<String, TickerInfos> = serde_json::from_str(&body)?;
+    Ok(TickerInfosRegistry(model))
+}
+
+pub async fn fetch_price_tickers(price_url: &str) -> Result<TickerInfosRegistry, MmError<PriceServiceRequestError>> {
+    let model = process_price_request(price_url).await?;
+    debug!("price registry size: {}", model.0.len());
+    Ok(model)
+}
+
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct CEXRates {
+    pub base: BigDecimal,
+    pub rel: BigDecimal,
+}
+
+// Fetcher funtion to test fetching latest prices from a endoint.
+async fn try_price_fetcher_endpoint(
+    endpoint: &str,
+    base: &str,
+    rel: &str,
+) -> Result<CEXRates, MmError<PriceServiceRequestError>> {
+    match process_price_request(endpoint).await {
+        Ok(response) => match response.get_cex_rates(base.to_string(), rel.to_string()) {
+            Some(fiat_price) => {
+                let (base_usd_price, rel_usd_price) = fiat_price.get_rate_price();
+                Ok(CEXRates {
+                    base: base_usd_price,
+                    rel: rel_usd_price,
+                })
+            },
+            None => {
+                let cex: Option<CEXRates> = None;
+                cex.or_mm_err(|| PriceServiceRequestError::Internal("Couldn't fetch price".to_string()))
+            },
+        },
+        Err(e) => MmError::err(e.into_inner()),
+    }
+}
+
+// Consume try_price_fetcher_endpoint result here using MY_PRICE_ENDPOINT1 and if it fails, we'll retry to fetch from MY_PRICE_ENDPOINT2.
+// Returning price data if successful or None if price fetching fails.
+pub async fn fetch_swap_coins_price(base: &str, rel: &str) -> Option<CEXRates> {
+    debug!("Trying to fetch coins latest price...");
+    for endpoint in [MY_PRICE_ENDPOINT1, MY_PRICE_ENDPOINT2] {
+        match try_price_fetcher_endpoint(endpoint, base, rel).await {
+            Ok(response) => return Some(response),
+            // Continue fetching endpoints.
+            Err(e) => error!("{:?}", e),
+        }
+    }
+    // Couldn't fetch prices.
+    None
+}
+
+#[cfg(test)]
+#[cfg(not(target_arch = "wasm32"))]
```

Comment:
Hmm, now I see that it can be simplified into:
```rust
#[cfg(all(test, not(target_arch = "wasm32")))]
```
Please look at the [example](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/conditional-compilation.html)